provider/mem/mem_test.go

```diff
@@ -72,6 +73,63 @@ func init() {
 	pretty.CompareConfig.IncludeUnexported = true
 }
 
+// TestAlertsSubscribePutStarvation tests starvation of `iterator.Close` and
+// `alerts.Put`. Both `Subscribe` and `Put` use the Alerts.mtx lock. `Subscribe`
+// needs it to subscribe and more importantly unsubscribe `Alerts.listeners`. `Put`
+// uses the lock to add additional alerts and iterate the `Alerts.listeners` map.
+// If the channel of a listener is at its limit, `alerts.Lock` is blocked, whereby
+// a listener can not unsubscribe as the lock is hold by `alerts.Lock`.
+func TestAlertsSubscribePutStarvation(t *testing.T) {
+	marker := types.NewMarker()
+	alerts, err := NewAlerts(marker, 30*time.Minute)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	iterator := alerts.Subscribe()
+
+	alertsToInsert := []*types.Alert{}
+	// Exhaust alert channel
+	for i := 0; i < alertChannelLength+1; i++ {
+		alertsToInsert = append(alertsToInsert, &types.Alert{
+			Alert: model.Alert{
+				// Make sure the fingerprints differ
+				Labels:       model.LabelSet{"iteration": model.LabelValue(strconv.Itoa(i))},
+				Annotations:  model.LabelSet{"foo": "bar"},
+				StartsAt:     t0,
+				EndsAt:       t1,
+				GeneratorURL: "http://example.com/prometheus",
+			},
+			UpdatedAt: t0,
+			Timeout:   false,
+		})
+	}
+
+	putIsDone := make(chan struct{})
+	putsErr := make(chan error, 1)
+	go func() {
+		if err := alerts.Put(alertsToInsert...); err != nil {
+			putsErr <- err
+			return
+		}
+
+		putIsDone <- struct{}{}
+	}()
+
+	// Increase probability that `iterator.Close` is called before `alerts.Put`
```

Comment:
super nit but missing dot. Also isn't it "`iterator.Close` is called **after** `alerts.Put`"?